### PR TITLE
NodeCG v0.9

### LINKED ISSUE #149: Replace more-routing with app-route once out of beta
[`carbon-route`](https://github.com/polymerelements/carbon-route) was just released as a beta. Once this element leaves beta, we should replace `more-routing` with it.

### PR DESCRIPTION
**Pull request by Lange**
This is not ready to merge yet.

Will be ready to merge once all issues on [milestone v0.9](https://github.com/nodecg/nodecg/issues?q=is%3Aopen+is%3Aissue+milestone%3Av0.9.0) are closed.

Closes #149.

### CODE REVIEW

**Comment by Lange** [comment]
This is potentially ready for release. I'm going to sit on it for a few more days as I do other work, just to see if any issues crop up.

File: package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "chokidar": "^1.4.2",
     "clone": "^2.0.0",
     "compression": "^1.6.0",
+    "convict": "^4.0.0",
```
Comment (haganbmj) [inline package.json]: Can `convict` remove `is-my-json-valid` from here?
https://github.com/nodecg/nodecg/blob/037709a9f78403b2d7a8e76b9487b48033fb4f14/lib/replicant/shared.js#L241
Looks to be the last usage in the project.

Same with `json-schema-defaults`
https://github.com/nodecg/nodecg/blob/775c15824b0ec4b4138c4a8677179549b0608fd5/lib/replicant/replicant.js#L113-L119

Comment (Lange) [reply package.json]: No, it cannot. Convict uses its own special schema format, which is not compatible (nor does it have the same goals/purpose) as JSON Schemas, which are used in replicants. Thank you for checking thought! The extra eyes looking for places to remove unnecessary packages are very appreciated.

Comment (haganbmj) [reply package.json]: Yep, wasn't entirely sure if it was a straight replacement for all schemas.

File: Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM node:6
+FROM node:7
```
Comment (haganbmj) [inline Dockerfile]: misc: travis is building/testing with node 8, docker is set to 7.

Comment (Lange) [reply Dockerfile]: good catch, fixed!

File: README.md [hunk 3]
```diff
@@ -1,35 +1,63 @@
...
+NodeCG is a broadcast graphics framework and application. It enables you to write complex, dynamic broadcast graphics
+using the web platform. NodeCG has no graphics or drawing primitives of its own. Instead, NodeCG provides
+a structure for your code and an API to facilitate moving data between the dashboard, the server, and your graphics.
+It makes no assumptions about how to best code a graphic, and gives you complete freedom to use whatever libraries, 
+frameworks, tools, and methodologies you want. As such, NodeCG graphics can be rendered in any environment that
+can render HTML, including:
+- [OBS Studio](https://obsproject.com/) (via the [obs-browser](https://github.com/kc5nra/obs-browser) plugin)
+- [vMix](http://www.vmix.com/)
+
+> Don't see your preferred streaming software on this list? NodeCG graphics require Chrome 49 or newer.
```
Comment (haganbmj) [inline README.md]: Predicting questions about obs classic/xsplit/casper support.
What're the chances babel/polyfill could resolve that at some point in the future. classic is eol, but the other two are pretty relevant.

Comment (Lange) [reply README.md]: The main browser feature that NodeCG now uses is [ES6 Proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), which cannot be polyfilled nor transpiled for. They are not possible to replicate in userspace JavaScript code -- they have to be implemented natively in the browser. ES Proxies were added to Chrome in version 49, so modern versions of NodeCG cannot run on older builds of Chrome.

XSplit actually does now support NodeCG graphics, as it currently uses a build of CEF based on Chrome 57. I will update the readme.

File: lib/api.js [hunk 4]
```diff
@@ -205,18 +267,31 @@ NodeCG.sendMessageToBundle = function (messageName, bundleName, data, cb) {
...
-			content: data
+		return new Promise((resolve, reject) => {
+			window.socket.emit('message', {
+				bundleName,
+				messageName,
+				content: data
+			}, (err, ...args) => {
+				if (typeof cb === 'function') {
+					cb(err, ...args);
+					return;
```
Comment (haganbmj) [inline lib/api.js]: Nice.
So this doesn't spit out UnhandledPromiseRejection notices, `return` just satisfies the promise resolution?

Comment (Lange) [reply lib/api.js]: It doesn't resolve nor reject the promise, which to my knowledge is totally okay.

Comment (Lange) [reply lib/api.js]: Maybe I should modify this to just not even return a promise if `cb` is a function.

Comment (Lange) [reply lib/api.js]: Pushed that change.

**Comment by Lange** [comment]
The last issue has been closed and a release is imminent. If anyone wants to get a review in before release, it'd be good to drop a comment here in the next 12 hours or so.

**Pull request merged**